src/activation.rs

```diff
@@ -0,0 +1,180 @@
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{wl_seat, wl_surface},
+    Dispatch, Proxy, QueueHandle,
+};
+use wayland_protocols::xdg::activation::v1::client::{xdg_activation_token_v1, xdg_activation_v1};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+#[derive(Debug, Clone)]
+pub struct RequestData {
+    pub app_id: Option<String>,
+    pub seat_and_serial: Option<(wl_seat::WlSeat, u32)>,
+    pub surface: Option<wl_surface::WlSurface>,
+}
+
+/// Data attached to a token request
+pub trait RequestDataExt: Send + Sync {
+    /// App_id of the application requesting the token, if applicable
+    fn app_id(&self) -> Option<&str>;
+    /// Seat and serial of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests without
+    /// recent serials. There is no way to detect this from the client-side.
+    fn seat_and_serial(&self) -> Option<(&wl_seat::WlSeat, u32)>;
+    /// Surface of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests from
+    /// unfocused surfaces. There is no way to detect this from the client-side.
+    fn surface(&self) -> Option<&wl_surface::WlSurface>;
+}
```

Comment:
The whole concept of that `with_data` methods is that you ask for `RequestData` from the `Ext` not for the fields, the fields are stored on the data you use inside the `SCTK`.

So all your docs should go to the `RequestData`, but this trait should just have `fn request_data(&self) -> &RequestData`. As you can see from the `KeyboardDataExt`.

Then you don't need to have associated type in user code as well iirc.

Reply:
> Then you don't need to have associated type in user code as well iirc.

I think I would still need to that to know what type the `RequestDataExt` is to return it in the `ActivationHandler`. The only thing that returning `RequestData` from the trait instead gives is easier composability, as the user may just store the `RequestData` in it's type. But they can already do that and forwarding three methods doesn't seem much less ergonomic to me.

Anyway, sure I can change that, but I don't see how this gets rid of the associated type.

Reply:
Ah, yes, it won't get it because you pass the `data`, SCTK usually passes the wayland object and user gets the data. It's a bit meh, but it's consistent.

Reply:
> SCTK usually passes the wayland object and user gets the data. It's a bit meh, but it's consistent.

Uhh, that feels entirely unnecessary given the user has never seen a token object up until this point.

Reply:
yeah, you can leave mostly as is, it's just sctk has not that great api with delegate and user data, it's very verbose.